system/View/Table.php

```diff
@@ -302,7 +302,9 @@ public function generate($tableData = null)
 
                 foreach ($heading as $key => $val) {
                     if ($key !== 'data') {
-                        $temp = str_replace('<th', '<th ' . $key . '="' . $val . '"', $temp);
+                        $headerElement = (bool) preg_match('/td/i', $temp) ? '<td' : '<th';
```

Comment:
It seems we do not need to call this  `preg_match()` every time.
And `'/<td/i'` is better?

Reply:
```php
preg_match('/td|th/i', $temp, $valueMatch);
$strVal = end($valueMatch);

$temp = $strVal ? str_replace($strVal, $strVal . ' ' . $key . '="' . $val . '"', $temp) : $temp;
```
What do you think above?

If do not need call `preg_match` every time,  in `generate()` set option ?

Reply:
The following test will fail.

```php
        $template = [
            'table_open'         => '<table border="1" cellpadding="4" cellspacing="0">',
            'thead_open'         => '<thead>',
            'thead_close'        => '</thead>',
            'heading_row_start'  => '<tr>',
            'heading_row_end'    => '</tr>',
            'heading_cell_start' => '<th>',
            'heading_cell_end'   => '</th>',
            'tfoot_open'         => '<tfoot>',
            'tfoot_close'        => '</tfoot>',
            'footing_row_start'  => '<tr>',
            'footing_row_end'    => '</tr>',
            'footing_cell_start' => '<td>',
            'footing_cell_end'   => '</td>',
            'tbody_open'         => '<tbody>',
            'tbody_close'        => '</tbody>',
            'row_start'          => '<tr>',
            'row_end'            => '</tr>',
            'cell_start'         => '<td>',
            'cell_end'           => '</td>',
            'row_alt_start'      => '<tr>',
            'row_alt_end'        => '</tr>',
            'cell_alt_start'     => '<td>',
            'cell_alt_end'       => '</td>',
            'table_close'        => '</table>',
        ];

        $this->table->setTemplate($template);
        $this->table->setHeading(['Name', ['data' => 'Amount', 'class' => 'tdk', 'style' => $this->styleTable]]);

        // ...
```

Reply:
![image](https://user-images.githubusercontent.com/97607754/189520675-147ffa60-7c6a-4fc1-bfd8-f417191228c9.png)

```php
    public function testGenerateWithClassStyle()
    {
        $template = [
            'table_open'         => '<table border="1" cellpadding="4" cellspacing="0">',
            'thead_open'         => '<thead>',
            'thead_close'        => '</thead>',
            'heading_row_start'  => '<tr>',
            'heading_row_end'    => '</tr>',
            'heading_cell_start' => '<th>',
            'heading_cell_end'   => '</th>',
            'tfoot_open'         => '<tfoot>',
            'tfoot_close'        => '</tfoot>',
            'footing_row_start'  => '<tr>',
            'footing_row_end'    => '</tr>',
            'footing_cell_start' => '<td>',
            'footing_cell_end'   => '</td>',
            'tbody_open'         => '<tbody>',
            'tbody_close'        => '</tbody>',
            'row_start'          => '<tr>',
            'row_end'            => '</tr>',
            'cell_start'         => '<td>',
            'cell_end'           => '</td>',
            'row_alt_start'      => '<tr>',
            'row_alt_end'        => '</tr>',
            'cell_alt_start'     => '<td>',
            'cell_alt_end'       => '</td>',
            'table_close'        => '</table>',
        ];

        $this->table->setTemplate($template);
        $this->table->setHeading(['Name', ['data' => 'Amount', 'class' => 'tdk', 'style' => $this->styleTable]]);

        $this->table->addRow(['Fred', 1]);
        $this->table->addRow(['Mary', 3]);
        $this->table->addRow(['John', 6]);

        $this->table->setFooting([['data' => 'Total', 'class' => 'tdk'], ['data' => '<small class="text-light">IDR <span class="badge badge-info">10</span></small>', 'style' => $this->styleTable]]);

        $table = $this->table->generate();

        // Header
        $this->assertStringContainsString('<th>Name</th>', $table);
        $this->assertStringContainsString('<th style="' . $this->styleTable . '" class="tdk">Amount</th>', $table);

        // Footer
        $this->assertStringContainsString('<td class="tdk">Total</td>', $table);
        $this->assertStringContainsString('<td style="' . $this->styleTable . '"><small class="text-light">IDR <span class="badge badge-info">10</span></small></td>', $table);
    }
```

With this PR condition using your suggest test fail

My [comment](https://github.com/codeigniter4/CodeIgniter4/pull/6511#discussion_r967587298) will return `td`, `th`, or `null`. If `null` will return default template. And success your suggest tests

Reply:
Neither code is good. Because if the tag is `<td>`, the start element is `td` and never change. 
So we should check it only once, and once is enough.

Reply:
How about now? `preg_match` is called by how many columns

@kenjis By the way, What coverall tool do you use in code editor?

Reply:
The screenshot is PhpStorm.
I use PhpStorm built in coverage reporting, or phpunit HTML report.